### PR TITLE
Add Smithy RPC v2 CBOR protocol tests

### PR DESCRIPTION
**Pull request by kstich**
#### Background
* What do these changes do? 
Add protocol tests for Smithy RPC v2 CBOR.
* Why are they important?
Enable protocol compliance testing for the protocol.

#### Testing
* How did you test these changes?
Validation in SDK implementations.

---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+        headers: {
+            "smithy-protocol": "rpc-v2-cbor",
+            "Accept": "application/cbor",
+            "Content-Type": "application/cbor"
+        }
+        method: "POST",
+        uri: "/service/RpcV2Protocol/operation/NoInputOutput",
+        body: "v/8=",
+        appliesTo: "server"
+    }
```
Comment (kstich) [inline smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy:68]: These test cases are the same.

File: smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+        },
+        forbidHeaders: [
+            "Content-Type"
+        ]
+        code: 200,
+    },
+    {
+        id: "no_output_client_allows_accept",
+        protocol: rpcv2Cbor,
+        documentation: "Servers should allow the accept header to be set to the default content-type.",
```
Comment (kstich) [inline smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy:88]: Says "Servers" but is a client test. I don't think `Accept` is relevant for responses either.

File: smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+        code: 200,
+        appliesTo: "client",
+    },
+    {
+        id: "no_input_client_allows_empty_cbor",
+        protocol: rpcv2Cbor,
+        documentation: "Client should accept CBOR empty struct if no output",
+        headers: {
+            "smithy-protocol": "rpc-v2-cbor",
+            "Accept": "application/cbor",
```
Comment (kstich) [inline smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy:104]: Not sure this is relevant in response.

File: smithy-aws-protocol-tests/model/rpcV2/cbor-lists.smithy [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+        params: {
+            sparseStringList: [null, "hi"]
+        }
+    }
+])
+
+structure RpcV2CborListInputOutput {
+    stringList: StringList,
+
+    sparseStringList: SparseStringList,
```
Comment (kstich) [inline smithy-aws-protocol-tests/model/rpcV2/cbor-lists.smithy:285]: Can all `@sparse` collections be split to their own operations, similar to #2206 ?

File: smithy-aws-protocol-tests/model/rpcV2/cbor-structs.smithy [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+        bodyMediaType: "application/cbor",
+        uri: "/service/RpcV2Protocol/operation/SimpleScalarProperties",
+        body: "oWlibG9iVmFsdWVfWCJBbiBleGFtcGxlIGluZGVmaW5pdGUtYnl0ZSBzdHJpbmcsUSBjaHVua2VkIG9uIGNvbW1h/w=="
+        params: {
+            blobValue: "An example indefinite-byte string, chunked on comma"
+        },
+        appliesTo: "server"
+    },
+    {
+        id: "RpcV2CborSupportsUpcastingData",
```
Comment (kstich) [inline smithy-aws-protocol-tests/model/rpcV2/cbor-structs.smithy:191]: Can there be a client side version of this for responses?

File: smithy-aws-protocol-tests/model/rpcV2/cbor-structs.smithy [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+            integerValue: 256,
+            shortValue: 9898,
+            stringValue: "simple",
+            blobValue: "foo"
+        },
+        appliesTo: "client"
+    },
+    {
+        id: "RpcV2CborClientDoesntDeSerializeNullStructureValues",
+        documentation: "RpcV2 Cbor should deserialize null structure values",
```
Comment (kstich) [inline smithy-aws-protocol-tests/model/rpcV2/cbor-structs.smithy:263]:
```suggestion
        id: "RpcV2CborClientDoesntDeserializeNullStructureValues",
        documentation: "RpcV2 Cbor should not deserialize null structure values",
```

**Review by kstich — COMMENTED** [review]
Should also add tags the tests that have definite/indefinite encodings so implementations that can't support one or the other for some reason can handle them.

File: smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy (hunk 2 above)
Comment (drganjoo) [reply smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy:88]: I agree, we shouldn't be testing `Accept` headers in response. Will delete this test.

File: smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy (hunk 1 above)
Comment (drganjoo) [reply smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy:68]: Fixing it.

File: smithy-aws-protocol-tests/model/rpcV2/cbor-lists.smithy [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+            "Content-Type": "application/cbor"
+        },
+        params: {
+            sparseStringList: [null, "hi"]
+        }
+    }
+])
+
+structure RpcV2CborListInputOutput {
+    stringList: StringList,
```
[the remark this answers is not in the document]
Comment (drganjoo) [reply smithy-aws-protocol-tests/model/rpcV2/cbor-lists.smithy:283]: Oh good catch. Adding it to the structure.

File: smithy-aws-protocol-tests/model/rpcV2/cbor-structs.smithy (hunk 5 above)
Comment (drganjoo) [reply smithy-aws-protocol-tests/model/rpcV2/cbor-structs.smithy:191]: Adding a test for the client.

File: smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy (hunk 3 above)
Comment (drganjoo) [reply smithy-aws-protocol-tests/model/rpcV2/empty-input-output.smithy:104]: Removing the accept header.

File: smithy-aws-protocol-tests/model/rpcV2/cbor-lists.smithy (hunk 4 above)
Comment (drganjoo) [reply smithy-aws-protocol-tests/model/rpcV2/cbor-lists.smithy:285]: They have been moved into separate operation.

**Comment by lucix-aws** [comment]
Passing in go v2

**Pull request merged**